src/lib.rs

```diff
@@ -15,5 +15,216 @@
+//! #         PROVIDER_NAME,
+//! #     ))),
+//! #     None,
+//! #     Some(Endpoint::immutable(Uri::from_static(MINIO_ENDPOINT))),
+//! #     None,
+//! #     None,
+//! #     None,
+//! # )
+//! # .await);
+//!
+//! let resolved_schema = listing_options
+//!     .infer_schema(s3_file_system.clone(), filename)
+//!     .await?;
+//!
+//! let table = ListingTable::new(
+//!     s3_file_system,
+//!     filename.to_owned(),
+//!     resolved_schema,
+//!     listing_options,
+//! );
+//!
+//! let mut ctx = ExecutionContext::new();
+//!
+//! ctx.register_table("tbl", Arc::new(table))?;
+//!
+//! let df = ctx.sql("SELECT * FROM tbl").await?;
+//! df.show();
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! We can also register the `S3FileSystem` directly as an `ObjectStore` on an `ExecutionContext`. This provides an idiomatic way of creating `TableProviders` that can be queried.
+//!
+//! ```rust
+//! use std::sync::Arc;
+//!
+//! use datafusion::datasource::listing::*;
+//! use datafusion::datasource::TableProvider;
+//! use datafusion::prelude::ExecutionContext;
+//! use datafusion::datasource::file_format::parquet::ParquetFormat;
+//! use datafusion::error::Result;
+//!
+//! use datafusion_objectstore_s3::object_store::s3::S3FileSystem;
+//!
+//! use aws_types::credentials::SharedCredentialsProvider;
+//! use aws_types::credentials::Credentials;
+//! use aws_sdk_s3::Endpoint;
+//! use http::Uri;
+//!
+//! # const MINIO_ACCESS_KEY_ID: &str = "AKIAIOSFODNN7EXAMPLE";
+//! # const MINIO_SECRET_ACCESS_KEY: &str = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY";
+//! # const PROVIDER_NAME: &str = "Static";
+//! # const MINIO_ENDPOINT: &str = "http://localhost:9000";
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<()> {
+//!
+//! # let s3_file_system = Arc::new(S3FileSystem::new(
+//! #     Some(SharedCredentialsProvider::new(Credentials::new(
+//! #         MINIO_ACCESS_KEY_ID,
+//! #         MINIO_SECRET_ACCESS_KEY,
+//! #         None,
+//! #         None,
+//! #         PROVIDER_NAME,
+//! #     ))),
+//! #     None,
+//! #     Some(Endpoint::immutable(Uri::from_static(MINIO_ENDPOINT))),
+//! #     None,
+//! #     None,
+//! #     None,
+//! # )
+//! # .await);
+//!
+//! let mut ctx = ExecutionContext::new();
+//!
+//! ctx.register_object_store("s3", s3_file_system.clone());
+//!
+//! let (object_store, name) = ctx.object_store("s3")?;
+//!
+//! let filename = "data/alltypes_plain.snappy.parquet";
+//!
+//! let listing_options = ListingOptions {
+//!     format: Arc::new(ParquetFormat::default()),
+//!     collect_stat: true,
+//!     file_extension: "parquet".to_owned(),
+//!     target_partitions: num_cpus::get(),
+//!     table_partition_cols: vec![],
+//! };
+//!
+//! let resolved_schema = listing_options
+//!     .infer_schema(s3_file_system.clone(), filename)
+//!     .await?;
+//!
+//! let table = ListingTable::new(
+//!     s3_file_system,
+//!     filename.to_owned(),
+//!     resolved_schema,
+//!     listing_options,
+//! );
```

Comment:
I cant get doc tests to pass using the `object_store` returned from `ctx.object_store` method (i get file not found error).  Does anything seem off to you @seddonm1?

Reply:
I cloned your repo, and ran the tests per the readme and they worked fine:

```
running 8 tests
test object_store::s3::tests::test_ctx_register_object_store ... ok
test object_store::s3::tests::test_read_nonexistent_bucket - should panic ... ok
test object_store::s3::tests::test_read_nonexistent_file ... ok
test object_store::s3::tests::test_read_alternative_bucket - should panic ... ok
test object_store::s3::tests::test_read_range ... ok
test object_store::s3::tests::test_read_parquet ... ok
test object_store::s3::tests::test_sql_query ... ok
test object_store::s3::tests::test_read_files ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 5.87s

   Doc-tests datafusion_objectstore_s3

running 4 tests
test src/lib.rs - (line 46) ... ok
test src/lib.rs - (line 31) ... ok
test src/lib.rs - (line 81) ... ok
test src/lib.rs - (line 153) ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 7.11s
```

Reply:
Thats because i updated it to use `s3_file_system` instead of `object_store` for now.  Locally, if you update to use `object_store` that is returned from `ctx.object_store("s3")` instead then i think it will fail.